primitives/src/campaign.rs

```diff
@@ -0,0 +1,239 @@
+use crate::{
+    channel_v5::Channel, targeting::Rules, AdUnit, BigNum, EventSubmission, ValidatorDesc,
+};
+
+use chrono::{
+    serde::{ts_milliseconds, ts_milliseconds_option},
+    DateTime, Utc,
+};
+use serde::{Deserialize, Serialize};
+
+pub use pricing::{Pricing, PricingBounds};
+pub use spec::{ValidatorRole, Validators};
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Campaign {
+    pub channel: Channel,
+    pub spec: CampaignSpec,
+}
+
+impl Campaign {
+    /// Matches the Channel.leader to the Campaign.spec.leader
+    /// If they match it returns `Some`, otherwise, it returns `None`
+    pub fn leader<'a>(&'a self) -> Option<&'a ValidatorDesc> {
+        if self.channel.leader == self.spec.validators.leader().id {
+            Some(self.spec.validators.leader())
+        } else {
+            None
+        }
+    }
+
+    /// Matches the Channel.follower to the Campaign.spec.follower
+    /// If they match it returns `Some`, otherwise, it returns `None`
+    pub fn follower<'a>(&'a self) -> Option<&'a ValidatorDesc> {
+        if self.channel.follower == self.spec.validators.follower().id {
+            Some(self.spec.validators.follower())
+        } else {
+            None
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct CampaignSpec {
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub title: Option<String>,
+    pub validators: Validators,
```

Comment:
Do we need to specify validators, we already have the Channel?

Reply:
@Ivshti I think it's safe to remove the validators array from the campaign spec. Maybe add `channel id` ?

Reply:
The v5 `Channel` only contains the addresses of the validators, the CampaignSpec validators include the `fee`, `fee_addr`, `fee_token`, `url`